### PR TITLE
Add comment in Customer registration form, add better error message

### LINKED ISSUE #13640: Better alert for customer creation and customer info edit forms
**Is your feature request related to a problem? Please describe.**
Following 1.7.5.2 patch, when an existing customer having a first name such as "John R.Hamond" wants to modify its account info an alert is raised when saving saying "Incorrect name".
This is because his/her name doesn't match the new forms' conditions.
Problem is that this name used to be good and is not anymore, we need to explain it.

Capture d’écran 2019-05-07 à 09 39 17

This alert is not explicit enough as the name was previously good and not anymore. We need to inform the customer about new conditions introduced in 1.7.5.2 .

**Todo:**
- [X] Study feasibilty
- [ ] Add new error message 

**Describe the solution you'd like**

It is first necessary to study the complexity to modify this text. The message might come from the database content validation and not from the form itself. In the first case, it would require to create a specific event.

**If the message is generated at the form level:**
An alert with more context/information about fields conditions for the "First name" and "Last name" fields:
"
Invalid name.
The following characters are not accepted: / \ ` ^ *
The characters . and 。 are allowed only if they are directly followed by a space or the end of the string.
"

**If the message is generated at the db level:**
Add a message below the fisrtname and lastname field. The message should always be displayed, like in this example: 
57281825-06116580-70ac-11e9-841d-acd7482c85fc

Text: 
"
not accepted: / \ ` ^ *
A space is necessary after a "."
"

---
Should be added to the account creation form, the edit account and the form in the checkout.

Thanks,

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Add comment in Customer registration form, add better error message
| Type?         | improvement
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #13640
| How to test?  | Try to register a new customer, you should see comments under Firstname and Lastname fields If trying to send an invalid value `John R.Hamond` a more detailed error message is displayed

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/14138)

### CODE REVIEW

File: classes/ValidateConstraintTranslator.php [hunk 1]
```diff
@@ -49,14 +49,32 @@ public function __construct(TranslatorInterface $translator)
...
                 'Shop.Forms.Errors'
             );
         }
 
+        if ($validator === 'isCustomerName') {
+            return $this->translator->trans(
+                'Invalid name',
+                [],
+                'Shop.Forms.Errors'
+            ) . "\n" .
```
Comment (PierreRambaud) [inline classes/ValidateConstraintTranslator.php]: use `PHP_EOL` instead

File: classes/form/CustomerFormatter.php [hunk 2]
```diff
@@ -133,6 +137,10 @@ public function getFormat()
                     'Shop.Forms.Labels'
                 )
             )
+            ->addAvailableValue(
+                'comment',
+                $this->translator->trans("Not accepted: / \ ` ^ *\nA space is necessary after a \".\"", array(), 'Shop.Forms.Help')
```
Comment (PierreRambaud) [inline classes/form/CustomerFormatter.php]:
```suggestion
                $this->translator->trans("Not accepted: / \ ` ^ *\nA space is necessary after a \".\"", [], 'Shop.Forms.Help')
```

File: classes/ValidateConstraintTranslator.php (hunk 1 above)
Comment (PierreRambaud) [inline classes/ValidateConstraintTranslator.php]: use `PHP_EOL` instead

File: themes/classic/templates/_partials/form-errors.tpl [hunk 3]
```diff
@@ -27,7 +27,7 @@
     {block name='form_errors'}
       <ul>
         {foreach $errors as $error}
-          <li class="alert alert-danger">{$error}</li>
+          <li class="alert alert-danger">{$error|strip_tags|nl2br nofilter}</li>
```
Comment (PierreRambaud) [inline themes/classic/templates/_partials/form-errors.tpl]: why strip_tags?

Comment (jolelievre) [reply themes/classic/templates/_partials/form-errors.tpl]: to be sure no html tag is inserted (since I use `nofilter`) only the converted `\n` remain

File: classes/ValidateConstraintTranslator.php (hunk 1 above)
Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php]: Done

File: classes/form/CustomerFormatter.php (hunk 2 above)
Comment (jolelievre) [reply classes/form/CustomerFormatter.php]: Done

File: themes/classic/templates/_partials/form-errors.tpl (hunk 3 above)
Comment (jolelievre) [reply themes/classic/templates/_partials/form-errors.tpl]: I can remove it, it will still work but I was worried of XSS injection 😛

Comment (PierreRambaud) [reply themes/classic/templates/_partials/form-errors.tpl]: only if someone include xss in translations, but we should be worried if it happens ^^

Comment (jolelievre) [reply themes/classic/templates/_partials/form-errors.tpl]: alright I'll remove it then

File: themes/classic/templates/_partials/form-fields.tpl [hunk 4]
```diff
@@ -170,7 +170,7 @@
           >
           {if isset($field.availableValues.comment)}
             <span class="form-control-comment">
-              {$field.availableValues.comment}
+              {$field.availableValues.comment|nl2br nofilter}
```
Comment (jolelievre) [inline themes/classic/templates/_partials/form-fields.tpl]: @PierreRambaud I also removed this `strip_tag`

File: themes/classic/templates/_partials/form-errors.tpl (hunk 3 above)
Comment (jolelievre) [reply themes/classic/templates/_partials/form-errors.tpl]: Done

**Review by PierreRambaud — DISMISSED** [review]
I don't like the concat but not other way from now :sob:

**Comment by jolelievre** [comment]
> I don't like the concat but not other way from now 😭

Me neither, but if I return an array the error blocks are separated.. This would require to change too many templates and error management..

File: classes/ValidateConstraintTranslator.php [hunk 5]
```diff
@@ -49,14 +49,32 @@ public function __construct(TranslatorInterface $translator)
...
         }
 
+        if ($validator === 'isCustomerName') {
+            return $this->translator->trans(
+                'Invalid name',
+                [],
+                'Shop.Forms.Errors'
+            ) . PHP_EOL .
+            $this->translator->trans(
+                'The following characters are not accepted: / \ ^ * `',
```
Comment (LouiseBonnard) [inline classes/ValidateConstraintTranslator.php]: I think we should tend towards shorter wordings, otherwise error notifications will be too present and discouraging. What about _Invalid characters: /\^*`_ instead?

File: classes/ValidateConstraintTranslator.php [hunk 6]
```diff
@@ -49,14 +49,32 @@ public function __construct(TranslatorInterface $translator)
...
+                [],
+                'Shop.Forms.Errors'
+            ) . PHP_EOL .
+            $this->translator->trans(
+                'The characters . and 。 are allowed only if they are directly followed by a space or the end of the string.',
```
Comment (LouiseBonnard) [inline classes/ValidateConstraintTranslator.php]: I wonder about the "。" character... it seems pretty specific, is it an absolute necessity to mention it?

File: classes/ValidateConstraintTranslator.php (hunk 5 above)
Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php]: Seems good to me, @samuel-pires wdyt?

File: classes/ValidateConstraintTranslator.php (hunk 6 above)
Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php]: Well this character is indeed forbidden, but maybe you're right If someone tries to use this character it is probably in order to use the breach so maybe it's not useful to indicate it for usual users

**Comment by marionf** [comment]
@jolelievre some returns, seen with @samuel-pires and @TristanLDD 

1. Delete the message under the fields

capture d'écran_1664

2. There are some other prohibited characters that are not accepted and not displayed in the message like: **0-9!<>,;?=+()@#"°{}_$%:** we should add them

3. Replace "The following characters are not accepted: " by "Invalid characters: "

4. Replace "The characters . and 。 are allowed only if they are directly followed by a space or the end of the string." by "A space is needed after "." and "。""

**Comment by LouiseBonnard** [comment]
@marionf, about the wording, you can also write _A space is required after "." and "。"_ to give a sense of obligation.

**Comment by jolelievre** [comment]
@marionf @LouiseBonnard @samuel-pires @TristanLDD 
All your modifications have been addressed, this should be good unless you find other issues?

File: classes/ValidateConstraintTranslator.php [hunk 7]
```diff
@@ -49,14 +49,32 @@ public function __construct(TranslatorInterface $translator)
...
+                [],
+                'Shop.Forms.Errors'
+            ) . PHP_EOL .
+            $this->translator->trans(
+                'Invalid characters: 0-9!<>,;?=+()@#"°{}_$%/\^*`',
+                [],
+                'Shop.Forms.Errors'
+            ) . PHP_EOL .
+            $this->translator->trans(
+                'A space is required after . and 。',
```
Comment (LouiseBonnard) [inline classes/ValidateConstraintTranslator.php:72]: Shouldn't it be written with quotation marks? Example: _A space is required after "." and "。"_

Comment (LouiseBonnard) [reply classes/ValidateConstraintTranslator.php:72]: I don't know what is the best practice.

Comment (ghost) [reply classes/ValidateConstraintTranslator.php:72]: Tristan was for the ""

Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php:72]: So you mean using `'A space is required after "." and "。"'`
(Isn't the space after 。 weird in this case?)

Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php:72]: NB: the space can't be removed, it's part of the character

Comment (ghost) [reply classes/ValidateConstraintTranslator.php:72]: @matks told us it is the equivalent of "." in China and that you can use it as a dot in a URL

Comment (ghost) [reply classes/ValidateConstraintTranslator.php:72]: But yeah you are right the space is part of it so I don't really understand how it can be used for a URL 🤔

Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php:72]: I think maybe some email browsers simplify them and convert them into regular `.` thus creating a clickable url
But my question was is it ok if we have this little space inside the `""` ?

Comment (LouiseBonnard) [reply classes/ValidateConstraintTranslator.php:72]: @jolelievre, as for me, it is!

Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php:72]: Great, then @PierreRambaud can you re-approve this PR please? ^^

**Review by matks — APPROVED** [review]

File: classes/ValidateConstraintTranslator.php (hunk 7 above)
Comment (matks) [reply classes/ValidateConstraintTranslator.php:72]: > I think maybe some email browsers simplify them and convert them into regular `.` thus creating a clickable url

I confirm, try visiting https://www.apple。com 😄

Comment (LouiseBonnard) [reply classes/ValidateConstraintTranslator.php:72]: But... this PR has been merged without the fix on the wording?

Comment (matks) [reply classes/ValidateConstraintTranslator.php:72]: > Great, then @PierreRambaud can you re-approve this PR please? ^^

@LouiseBonnard we've been tricked !

Comment (jolelievre) [reply classes/ValidateConstraintTranslator.php:72]: Oups my bad, I was too quick and thought I had already added it
Here's a PR to fix this ^^
https://github.com/PrestaShop/PrestaShop/pull/14163
